external-crates/move/crates/move-package/src/resolution/dependency_graph.rs

```diff
@@ -1066,70 +1066,72 @@ impl DependencyGraph {
+            name: dep_name,
             subst,
             digest,
         } in packages.root_dependencies.into_iter().flatten()
         {
             package_graph.add_edge(
                 root_package_id,
-                Symbol::from(name.as_str()),
+                PackageIdentifier::from(dep_id.as_str()),
                 Dependency {
                     mode: DependencyMode::Always,
                     subst: subst.map(parse_substitution).transpose()?,
                     digest: digest.map(Symbol::from),
                     dep_override: false,
-                    dep_name: PM::PackageName::from(name),
+                    dep_name: PM::PackageName::from(dep_name),
                 },
             );
         }
 
         for schema::Dependency {
-            name,
+            id: dep_id,
+            name: dep_name,
             subst,
             digest,
         } in packages.root_dev_dependencies.into_iter().flatten()
         {
             package_graph.add_edge(
                 root_package_id,
-                Symbol::from(name.as_str()),
+                PackageIdentifier::from(dep_id.as_str()),
                 Dependency {
                     mode: DependencyMode::DevOnly,
                     subst: subst.map(parse_substitution).transpose()?,
                     digest: digest.map(Symbol::from),
                     dep_override: false,
-                    dep_name: PM::PackageName::from(name.as_str()),
+                    dep_name: PM::PackageName::from(dep_name),
                 },
             );
         }
 
         // Fill in the remaining dependencies, and the package source information from the lock
         // file.
         for schema::Package {
-            name: pkg_name,
+            id: pkg_id,
             source,
             version,
             dependencies,
             dev_dependencies,
         } in packages.packages.into_iter().flatten()
         {
-            let pkg_name = PM::PackageName::from(pkg_name.as_str());
-            let source = parse_dependency(pkg_name.as_str(), source)
-                .with_context(|| format!("Deserializing dependency '{pkg_name}'"))?;
+            let pkg_id = PackageIdentifier::from(pkg_id.as_str());
+            let source = parse_dependency(pkg_id.as_str(), source)
+                .with_context(|| format!("Deserializing dependency '{pkg_id}'"))?;
 
             let source = match source {
                 PM::Dependency::Internal(source) => source,
                 PM::Dependency::External(resolver) => {
-                    bail!("Unexpected dependency '{pkg_name}' resolved externally by '{resolver}'");
+                    bail!("Unexpected dependency '{pkg_id}' resolved externally by '{resolver}'");
                 }
             };
 
             if source.subst.is_some() {
-                bail!("Unexpected 'addr_subst' in source for '{pkg_name}'")
+                bail!("Unexpected 'addr_subst' in source for '{pkg_id}'")
```

Comment:
This is not super important, given it's lock file parsing, so if it goes wrong it's really more of an internal error, but I noticed that these are using the `pkg_id` rather than a name -- can we/should we use the `name` from its `source` here (and in other cases below) instead (or use both designators?)

Reply:
Well spotted, I forgot to mention this. I think using `name` from the manifest could make sense here in principle but it's a bit tricky in practice because package's name isn't available from the `[move.package]` section, but only as edges in the `dependencies` field list. The `name` field was put there (instead of the `[move.package]` section) because this is more natural from the dependency graph code and also it makes sense because in principle the same package could have a different `name` in case it was loaded from a different location (with a different name), or e.g. as both a source and an on-chain dep. So using a unique `name` for each unique package could lead to confusing error messages again.

Though in theory it would be possible to figure out the `name` for these error messages based on the graph, it would be a bit tricky to implement and I think using `pkg_id` makes sense anyways as these are errors in the lockfile so referencing ids is ok (as this is what you'd be looking for anyways if debugging).

Reply:
I initially thought this would be simpler because we had `source` which is a `PM::Dependency` lying around, and I got it mixed up with the lock file's `schema::Dependency` which had the name added to it -- so fair enough that this is currently too tricky to reconcile.

> this is more natural from the dependency graph code and also it makes sense because in principle the same package could have a different name in case it was loaded from a different location (with a different name), or e.g. as both a source and an on-chain dep. So using a unique name for each unique package could lead to confusing error messages again.

This raises an interesting point though, because prior to this change, we did have a rule that packages must have the same name (because we check that the name it's referred to when mentioned as a dependency has to match its package name). Do we still have that check? i.e. if we have a dependency:

```toml
Foo = { git = "..." }
```

Do we check that the package we are pointing to is called `Foo` (because it's not an on-chain dependency in this case)? This is separate from the question of whether we can try and recover a single unique package name for an on-chain package though, because I don't think we want to carry this requirement over to on-chain only deps.

Reply:
> we did have a rule that packages must have the same name (because we check that the name it's referred to when mentioned as a dependency has to match its package name)

We still have that check, here's more context https://github.com/MystenLabs/sui/pull/15315/files#r1430206985. There is actually a unit test for this in `move-package`.